package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,579 @@
+                                    npairs += 1
+        return results
+
+
+    def self_search(self):
+        cdef ns_int i, j, size_search
+        cdef ns_int d, m
+        cdef ns_int current_beadid, bid
+        cdef ns_int cellindex, cellindex_probe
+        cdef ns_int xi, yi, zi
+
+
+        cdef NSResults results
+
+        cdef ns_int size = self.coords_bbox.shape[0]
+        cdef ns_int[:] checked = np.zeros(size, dtype=np.int)
+
+        cdef real d2
+        cdef rvec probe
+
+        cdef real cutoff2 = self.cutoff * self.cutoff
+        cdef ns_int npairs = 0
+
+        size_search = self.coords.shape[0]
+
+        results = NSResults(self.cutoff, self.coords, self.coords)
+
+        with nogil:
+            for i in range(size_search):
+                # Start with first search coordinate
+                current_beadid = i
+                # find the cellindex of the coordinate
+                cellindex = self.grid.cellids[current_beadid]
+                for xi in range(DIM):
+                    for yi in range(DIM): 
+                        for zi in range(DIM):
+                            # Calculate and/or reinitialize shifted coordinates
+                            #Probe the search coordinates in a brick shaped box
+                            probe[XX] = self.coords_bbox[current_beadid, XX] + (xi - 1) * self.grid.cellsize[XX]
+                            probe[YY] = self.coords_bbox[current_beadid, YY] + (yi - 1) * self.grid.cellsize[XX]
+                            probe[ZZ] = self.coords_bbox[current_beadid, ZZ] + (zi - 1) * self.grid.cellsize[XX]
+                            # Make sure the shifted coordinates is inside the brick-shaped box
+                            for m in range(DIM - 1, -1, -1):
+                                while probe[m] < 0:
+                                    for d in range(m+1):
+                                        probe[d] += self.box.c_pbcbox.box[m][d]
+                                while probe[m] >= self.box.c_pbcbox.box[m][m]:
+                                    for d in range(m+1):
+                                        probe[d] -= self.box.c_pbcbox.box[m][d]
+
+                            # Get the cell index corresponding to the probe
+                            cellindex_probe = self.grid.coord2cellid(probe)
+                            #for this cellindex search in grid
+                            for j in range(self.grid.nbeads[cellindex_probe]):
+                                bid = self.grid.beadids[cellindex_probe * self.grid.nbeads_per_cell + j]
+                                if bid < current_beadid:
+                                    continue
+                                #find distance between search coords[i] and coords[bid]
+                                d2 = self.box.fast_distance2(&self.coords_bbox[current_beadid, XX], &self.coords_bbox[bid, XX])
+                                if d2 < cutoff2:

```

Comment:
looks like npairs has the wrong indentation here, you don't want to increment if `d2 < EPSILON`.  I'd do `if d2 cutoff and d2 > EPSILON:`